bftcosi/bftcosi.go

```diff
@@ -210,59 +217,84 @@ func (bft *ProtocolBFTCoSi) Dispatch() error {
+			}
+			// Finish the prepare round
+			if err := bft.handleChallengePrepare(c); err != nil {
+				return err
+			}
+			if !bft.IsLeaf() {
+				if err := bft.handleResponsePrepare(bft.responseChan); err != nil {
+					return err
+				}
+			}
+			gotChallengePrepare = true
+		case c := <-bft.challengeCommitChan:
 
-	// Finish the commit round
-	if err := bft.handleChallengeCommit(<-bft.challengeCommitChan); err != nil {
-		return err
-	}
-	if !bft.IsLeaf() {
-		if err := bft.handleResponseCommit(bft.responseChan); err != nil {
-			return err
+			// Finish the commit round
+			if err := bft.handleChallengeCommit(c); err != nil {
+				return err
+			}
+			if !bft.IsLeaf() {
+				if err := bft.handleResponseCommit(bft.responseChan); err != nil {
+					return err
+				}
+			}
+			bft.successfulMut.Lock()
+			bft.successful = true
+			bft.successfulMut.Unlock()
+			// we return here even when prepare is not done yet,
+			// because we may not have received the challenge for
+			// the prepare phase
+			return nil
+		case <-time.After(bft.Timeout):
+			return fmt.Errorf("timeout waiting for challenge - " +
+				"might be OK because the protocol already finished")
 		}
 	}
-	return nil
 }
 
 // Signature will generate the final signature, the output of the BFTCoSi
-// protocol.
-// The signature contains the commit round signature, with the message.
-// If the prepare phase failed, the signature will be nil and the Exceptions
-// will contain the exception from the prepare phase. It can be useful to see
-// which cosigners refused to sign (each exceptions contains the index of a
-// refusing-to-sign signer).
-// Expect this function to have an undefined behavior when called from a
-// non-root Node.
+// protocol.  The signature contains the commit round signature, with the
+// message.  Expect this function to have an undefined behavior when called
+// from a non-root Node.  If the root does not finish correctly, then the
+// signature will be nil.
 func (bft *ProtocolBFTCoSi) Signature() *BFTSignature {
-	bftSig := &BFTSignature{
-		Sig:        bft.commit.Signature(),
-		Msg:        bft.Msg,
-		Exceptions: nil,
-	}
-	if bft.signRefusal {
-		bftSig.Sig = nil
-		bftSig.Exceptions = bft.tempExceptions
+
+	bft.successfulMut.Lock()
+	ok := bft.successful
+	bft.successfulMut.Unlock()
+	if !ok {
+		return &BFTSignature{
+			Msg: bft.Msg,
+		}
 	}
 
-	// This is a hack to include exceptions which are the result of offline
-	// nodes rather than nodes that refused to sign.
-	if bftSig.Exceptions == nil {
-		for _, ex := range bft.tempExceptions {
-			if ex.Commitment.Equal(bft.Suite().Point().Null()) {
-				bftSig.Exceptions = append(bftSig.Exceptions, ex)
+	// create exceptions for the nodes that were late to respond in the
+	// commit phase
+	var i int
+	exceptions := make([]Exception, bft.AllowedExceptions)
+	for _, tn := range bft.Children() {
+		if _, ok := bft.respondedNodesCommit[tn.ServerIdentity.Public]; !ok {
+			exceptions[i] = Exception{
```

Comment:
With regard to the error we saw - are we guaranteed here that the following holds true? `i < bft.AllowedExceptions`?

Reply:
I can't provide a 100% guarantee, but it should. The only way for it to be false would be a programmer error, which should make the code panic and then we need to fix the bug.